### DISCUSSION TITLE
tctl must be either used on the auth server (on the auth server)

### DISCUSSION THREAD

**Discussion by cwhiting-fusable** [gravitational/teleport#22110] [Q&A]
Hi there!

I was following the tutorial for AWS terraform: https://goteleport.com/docs/deploy-a-cluster/deployments/aws-terraform/, and had logged into one of the auth servers (via the bastion), and run:
`sudo tctl users add teleport-admin --roles=editor,access --logins=root`
which returns:
"tctl must be either used on the auth server or provided with the identity file via --identity flag"

Based on `journalctl -u teleport-auth` I noticed the launch configuration (which is being depreciated in favor of launch templates, FYI) started a bunch of EC2 instances without metadata/instance tags access: "Could not fetch EC2 instance's tags, please ensure 'allow instance tags in metadata' is enabled on the instance. labels/cloud.go:140", so I changed that in the ec2 settings and restarted the service with systemd.

The teleport nodes still don't think they're the main nodes.

What do I do to debug / repair this?

**Comment by stevenGravy** [comment]
To use `tctl` not on the auth server you must log in via `tsh login` with a role that has access to the `user` resources.  There is no relationship between the nodes and `tctl` for the non auth services.

**Reply by cwhiting-fusable** [reply]
per the instructions, it says to log into the auth node to run a few commands to create the first user. When I'm in the auth EC2 ASG node, it doesn't believe I'm in the auth node.

**Reply by stevenGravy** [reply]
Thanks for clarifying.  Is your `teleport.yaml` at `/etc/teleport.yaml` or another location?  Please add that in `-c <path>/teleport.yaml` if in another location as part of your `tctl` command.

**Reply by cwhiting-fusable** [reply]
so you're saying at this step: https://goteleport.com/docs/deploy-a-cluster/deployments/aws-terraform/#accessing-the-cluster-after-terraform-setup the docs should say `sudo tctl users add teleport-admin --roles=editor,access --logins=root -c /etc/teleport.yaml`?

I'm not positive how the official AMI's work, it looks like they're setting some user data: https://github.com/gravitational/teleport/blob/master/examples/aws/terraform/ha-autoscale-cluster/auth-user-data.tpl - so maybe it writes /etc/teleport.yaml when it's done. I'd have to log in again to check.

**Reply by stevenGravy** [reply]
If you are doing this on the auth node it's at `/etc/teleport.yaml` and `sudo tctl` should work.  Can you confirm its running the auth service?

`sudo service teleport-auth status`

**Reply by webvictim** [reply]
@RRcwhiting For context, the AMIs do run a systemd unit called `teleport-generate-config` which will generate and write `/etc/teleport.yaml` for you based on the contents of `/etc/teleport.d/conf`. That file is in turn generated by the instance's user-data, which gets configured via Terraform as you noted.

The script is here: https://github.com/gravitational/teleport/blob/master/assets/aws/files/bin/teleport-generate-config

If `/etc/teleport.yaml` doesn't exist on EC2 instances in your auth server ASG, it probably means that the script failed to run for some reason. The output of `sudo systemctl status teleport-generate-config`  and `sudo systemctl status teleport-auth` will probably help to figure out why that is.

**Reply by tiago-taveira** [reply]
Hi there, I'm facing the same issue and if I run `sudo service teleport-auth status` it results in:
```
Redirecting to /bin/systemctl status teleport-auth.service
● teleport-auth.service - Teleport Auth Service
   Loaded: loaded (/etc/systemd/system/teleport-auth.service; enabled; vendor preset: disabled)
   Active: activating (auto-restart) (Result: exit-code) since Wed 2023-04-12 13:39:46 UTC; 4s ago
  Process: 6288 ExecStart=/usr/local/bin/teleport start --config=/etc/teleport.yaml --diag-addr=127.0.0.1:3000 --pid-file=/run/teleport/teleport.pid (code=exited, status=1/FAILURE)
 Main PID: 6288 (code=exited, status=1/FAILURE)

Apr 12 13:39:46 ip-172-31-1-209.eu-central-1.compute.internal systemd[1]: Unit teleport-auth.service entered failed state.
Apr 12 13:39:46 ip-172-31-1-209.eu-central-1.compute.internal systemd[1]: teleport-auth.service failed.
```

do you have any guidance on how to fix this?

**Reply by webvictim** [reply]
What does `sudo journalctl --no-pager -u teleport-auth` say? There will probably be a more specific error there.

**Reply by mighty-services** [reply]
Hy There,

I´m having the same error output. I get `ERROR: tctl must be either used on the auth server or provided with the identity file via --identity flag` while following along the setup of ansible in my selfhosted environment. At the end of the [Prerequisites of the tutorial](https://goteleport.com/docs/machine-id/guides/ansible/#prerequisites) I should login to the teleport cluster from my ansible machine and issue `tctl status`. There it gives me the warning:
![grafik](https://github.com/gravitational/teleport/assets/71066192/fcd76b33-8f20-4be1-bde1-da65ffd0f7dd)

I have Teleport 14.0.0 on my self-hosted Cluster and 14.0.1 on my Ansible machine. The login worked, as you can see with tsh:
![grafik](https://github.com/gravitational/teleport/assets/71066192/e6121bb9-3aeb-48c7-b05b-f2c76f4e0f2e)

**Reply by webvictim** [reply]
Do you have an `/etc/teleport.yaml` file on this machine? Try running `TELEPORT_CONFIG_FILE="" tctl status` instead.

**Reply by cwhiting-fusable** [reply]
shouldn't... functionally... the ansible (or indeed autoscaling group) instructions given by Gravitational be complete / corrected, if /etc/teleport.yaml is a requirement? or that the script is in some way wrong / not lining up with the testing process and actual environments?

**Reply by webvictim** [reply]
The user who replied here is not running `tctl` on the Teleport auth server, they're attempting to run it remotely (see the hostname in the screenshots). This is a different situation to the issue that you originally posted.

Running `tctl` remotely won't work if the user (coincidentally) has an `/etc/teleport.yaml` config file present on the same server where they're running `tctl`, as `tctl` will then assume it is being run on the auth server and not remotely.

The easiest way to work around this fact is to `export TELEPORT_CONFIG_FILE=""` to tell `tctl` not to read that file - the alternatives are to delete the `/etc/teleport.yaml` file or run `tctl` on a machine which doesn't have an `/etc/teleport.yaml` file, both of which are difficult to do quickly and may result in unintended consequences.

Yes, it'd definitely be better if `tctl` could more accurately detect whether there is a Teleport auth server running locally or not, but this involves needing file permissions to read the `data_dir`/`/var/lib/teleport` directory (which it usually doesn't have when running as a non-root user). `tctl` must also remain backward compatible with older deployments of Teleport, meaning that the behaviour of reading the `/etc/teleport.yaml` file if present must be preserved.

TL;DR It's harder than it sounds, and we don't control where people run `tctl`!

**Reply by mighty-services** [reply]
Oh, I'm sorry, I didn't read the docs properly. I have to issue the tctl command on the cluster not on the ansbile-node. There it's working - I'm sorry, that's my fault!
